cluster: allow version bump without naming the version

tbg opened the pull request:
This simplifies documentation and usage, especially when it comes to
alpha releases.

I'm accepting bikesheds on the name of the alias.

Fixes https://github.com/cockroachdb/docs/issues/2213.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20292)
<!-- Reviewable:end -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/settings/cluster/settings.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292#-L--Hz6pofpddorYEVs0:-L--Hz6pofpddorYEVs1:blm668z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c17d4115c0503911f4779e5828b22898720b54d/pkg/settings/cluster/settings.go#L320)):*
> ```Go
> 	// We have a new proposed update to the value, validate it.
> 	var minVersion roachpb.Version
> 	if *versionBump == "from-binary" {
> ```

`current`? `latest`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/settings/cluster/settings.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292#-L--Hz6pofpddorYEVs0:-L--InvkiMABd0hP3Np1:biq5m7s) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c17d4115c0503911f4779e5828b22898720b54d/pkg/settings/cluster/settings.go#L320)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

`current`? `latest`?
</blockquote></details>

Those were my initial thoughts too, but `latest` is confusing (say you're updating from 1.1 to 1.2 but actually 1.3 has already been released; you'd wonder what "latest" would mean). Similar issue for `current`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/settings/cluster/settings.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292#-L--Hz6pofpddorYEVs0:-L--JcekTX8zYLCE4Sdx:b-4bh9pu) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c17d4115c0503911f4779e5828b22898720b54d/pkg/settings/cluster/settings.go#L320)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Those were my initial thoughts too, but `latest` is confusing (say you're updating from 1.1 to 1.2 but actually 1.3 has already been released; you'd wonder what "latest" would mean). Similar issue for `current`.
</blockquote></details>

Good points. Not too fond of `from-binary`, but it does have some upsides.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292)*
<!-- Sent from Reviewable.io -->

a-robinson reviewed (approved):
I like having this option, and `from-binary` seems reasonable to me. It'd be nice if there was an easier way to check what the `from-binary` version was before blindly using it, though. Perhaps only having it for stable releases is enough, though, and we can just point to `SELECT version()`.

benesch commented:
Drive by suggestion: introduce a function called `latest_known_version()` or similar, and then have the docs suggest:

```
SET CLUSTER SETTING version = latest_known_version()
```

Then it's hopefully obvious that you can use `SELECT latest_known_version()` to figure out what version you're about to upgrade to.

jseldess commented:
@benesch, wouldn't you want a function that tells you the version number specifically for the binary you upgraded to, not necessarily the latest?

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/settings/cluster/settings.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292#-L--Hz6pofpddorYEVs0:-L-1AjSzSAHogwyk2hXa:bh9dkjg) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c17d4115c0503911f4779e5828b22898720b54d/pkg/settings/cluster/settings.go#L320)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Good points. Not too fond of `from-binary`, but it does have some upsides.
</blockquote></details>

Yeah, I think it's important that we make the connection between the setting name and the binary that was just upgraded to, as opposed to "latest" or "current" in the abstract. What about `binary-version`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292#-:-L-1ARIbZX-fQaNlcbPy:bpjci0a)*
<!-- Sent from Reviewable.io -->

knz commented:
I am a bit confused about the desired semantics of this new feature. If there are multiple nodes in the clusters with different versions, the proposed implementation here will cause an upgrade to different version numbers *depending on which node processes the SET CLUSTER SETTING statement*. Obviously this won't do well especially if the user uses a load balancer.

I can see two alternatives:

- clarify in the syntax which node the latest version is to be pulled from. For example `set cluster setting version = current_node_version(3)` to pull the version from (e.g.) node 3
- make the semantics pick the highest version available across all nodes. For example `set cluster setting version = 'highest-supported-version'`

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/settings/cluster/settings.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292#-L--Hz6pofpddorYEVs0:-L-1Cjlz2YgSu9F4jITR:b-e4ytew) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c17d4115c0503911f4779e5828b22898720b54d/pkg/settings/cluster/settings.go#L320)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Yeah, I think it's important that we make the connection between the setting name and the binary that was just upgraded to, as opposed to "latest" or "current" in the abstract. What about `binary-version`?
</blockquote></details>

The word "binary" is a bit unclear IMHO. After all "binary" is just a numbering system, so "from-binary" would seem to suggest the version number needs to be decoded from 1001010101 or who knows what.

This is clearer and explicit: `current-executable-version`

(the discussion at the top of my review comment notwithstanding)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292#-:-L-1DHy3qxEDK-zvag80:b-gi4b03)*
<!-- Sent from Reviewable.io -->

jseldess commented:
@knz, my understanding is that you only need to bump the cluster version when upgrading the entire cluster to a new major version (e.g., 1.1.x to 1.2). At least that's what we say in the rolling upgrade docs. We also say you shouldn't do this until all nodes are on the same new version, so would a `highest-supported-version`scenario ever apply?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292#-:-L-23XdV9C24C_ardr77:b-xh7lj2)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@knz as @jseldess pointed out, if you're running a mixed version cluster and are trying to issue cluster version bumps, you can shoot yourself in the foot today. I like @benesch's idea of `latest_known_version` (which looks like it's similar to @knz's idea), but won't implement that myself. I'm happy to turn over this PR to anyone inclined to follow through.

Otherwise, as an alternative, we can expose the `ServerVersion` via SQL (don't think you can get it from anywhere right now; `version()` doesn't have the right thing for alphas) and make querying it part of the instructions, but I don't see how that is better than (a suitable renaming in) this code.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20292#-:-L-2DPT_PX8FgjHQfvk0:bsv1izr)*
<!-- Sent from Reviewable.io -->

knz commented:
What about populating that version number in `crdb_internal.node_build_info` with the other details?

tschottdorf commented:
@knz I don't want to tell people to run a random query on `crdb_internal` to get the right thing, but I'd be amenable to a top-level method as suggested by @benesch. `node_{binary,executable,running}_version()`?

benesch commented:
Ooh, I like `node_executable_version`.

knz commented:
place it in namespace `crdb_internal` still (just name the builtin with the string `"crdb_internal.node_executable_version"`) to outline that we're not yet 100% sure about the interface?

tschottdorf commented:
@knz done.

Ready for another look.

tschottdorf commented:
@jseldess, I assume you're happy with this, right?

jseldess reviewed (approved):
Yep. LGTM.

Just so I'm clear, now that we have this function, we will want to recommend using it whenever upgrading from one major version to another, since it simplifies things?

tschottdorf commented:
@jseldess yes, the docs would simply suggest

1. connect to one of the nodes (since you upgraded every node in the cluster, they are now all running the same, new, version)
2. `SELECT crdb_internal.node_executable_version()` and check that it matches your expectations
3. `SET CLUSTER SETTING version = crdb_internal.node_executable_version()`.

The upshot is that you wouldn't ever have to mention the specific version in this part of the docs, and it would work equally well for alphas (where `crdb_internal.node_executable_version()` could be, for example, `1.1-4`).

tschottdorf commented:
@RaduBerinde as resident plumbing expert, could you take a look? Had some little quarrels with `EvalContext`.

tbg committed:
cluster: add crdb_internal.node_executable_version()

This simplifies documentation and usage, especially when it comes to
alpha releases.

Fixes https://github.com/cockroachdb/docs/issues/2213.

Release note (sql change): Simplify cluster upgrades by providing
crdb_internal.node_executable_version().
